### PR TITLE
[tt-train] Qwen3 32B FSDP GRPO Bringup

### PR DESCRIPTION
**Pull request by awliu-TT**
### PR Category
This PR is for the bringup of Qwen3 32B, which requires lazy init + FSDP integration in GRPO.

### Summary
Things that changed in this PR:

1. A Qwen completer and helper files were added to enable lazy init + FSDP training of Qwen3 32B in GRPO.
2. GRPO trainer gains FSDP detection (named "fsdp" axis) and `ttml.sync_gradients` over ("dp","fsdp").
3. **[Bug Fix]**: A bug in the RMSNorm python op (which Qwen uses) was fixed. Originally, the RMSNorm python op was saving the Tensor buffer for use during the backwards pass. However, with FSDP and `reshard_after_forward`, the buffer becomes stale. The fix is to save the Tensor wrapper object and retrieve the buffer the object points to during the backwards pass so that we are always referencing a non-stale buffer.
4. Added WandB logging to the GRPO trainer.
5. Added a device config param for using lazy init. Defaults to true.
6. Added early checks in `GRPOConfig.__post_init__` to error out GRPO training if any of `num_generations`, `gradient_accumulation_steps`, `per_device_train_batch_size`, `prompts_to_train`, and (if checkpointing enabled) `checkpoint_interval` are <= 0.
7. Updated the `GRPO_trainer.md` with Qwen3 32B and FSDP info.
8. Reorganized Qwen3 code that was shared between the example `examples/qwen3/utils` and GRPO to avoid code duplication.
  a. Moved the shared APIs into importable modules `tt-train/sources/ttml/ttml/models/qwen3/weights.py` (APIs for loading HF weights imported from `ttml.models.qwen3.weights`) and `tt-train/sources/ttml/ttml/models/qwen3/kv_cache.py` (Python KV cache wrapper imported from `ttml.models.qwen3.kv_cache`).
  b. Moved the host to device transfer of weights in `load_weights_from_hf` and `load_weights_from_hfload_weights_from_hf_distributed`  from the concurrent threads to the main thread (serialized this part). Device ops are not thread safe, and running those device ops concurrently across worker threads races on the program-cache binary commit (`TT_FATAL "Expected Program Binaries to be committed to DRAM"`). Host-side preparation of the weights (float cast, permutation, padding) is still done multithreaded.
  c. Moved the API `build_causal_mask` from `ttml.common.data` to `ttml.common.utils` and added an argument to return a device tensor (defaults to false) to replace the `create_causal_mask` function previously defined in both the `grpo` and `examples/qwen3/utils` folders.
  d. Created the API `build_mesh` in `tt-train/sources/ttml/ttml/common/utils.py` to replace the identical APIs previously defined in both the GRPO Qwen3 utils and in `tt-train/sources/examples/train/train.py`.
8. **[Bug Fix]**: In `tt-train/sources/ttml/ttml/common/config.py` `TransformerConfig`, `self.head` should be set to `None` when no `head_dim` field is provided, instead of trying to set `int(None)`, which throws an error. **UPDATE** [Kevin's PR fixed this](https://github.com/tenstorrent/tt-metal/pull/48272).
9. **[Bug Fix]**: (Not related to the rest of my changes, but happened to be discovered while doing the API migrations) In `tt-train/sources/examples/python/transformers/training.py`, we were importing `get_config` from `ttml.common.config`. This API was changed and renamed to `load_config` in f0eece2693d2935b42d36f7a1191d5b6e711ae9e.

Beyond the scope of this PR:

1. Checkpointing (to be integrated next).
2. Perf optimizations (to be explored next).

Here is the rewards curve from a boolq GRPO run with Qwen3 32B:
image

- [x] [Sanity Tests](https://github.com/tenstorrent/tt-metal/actions/runs/28224221619)
- [x] [Blackhole Post Commit](https://github.com/tenstorrent/tt-metal/actions/runs/28224243686)
- [x] [Nightly L2](https://github.com/tenstorrent/tt-metal/actions/runs/28264301415) - [main results](https://github.com/tenstorrent/tt-metal/actions/runs/28251017287/job/83710295507)
- [x] [(Runtime) Sanity Tests](https://github.com/tenstorrent/tt-metal/actions/runs/28224316068)
- [x] New/Existing unit tests for this change

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:awliu/qwen_3_32B_FSDP_grpo)

### CODE REVIEW

File: tt-train/sources/examples/grpo/boolq_training_example.py [hunk 1]
```diff
@@ -41,23 +61,151 @@ def on_step_end(self, trainer, step, **kwargs):
...
+
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+
+    is_qwen3 = args.model == "qwen3"
+
+    if is_qwen3:
+        model_id = args.model_source or "Qwen/Qwen3-0.6B"
+        default_config = "tt-train/configs/training_configs/grpo_boolq_qwen3_fsdp_validate.yaml"
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/examples/grpo/boolq_training_example.py]: Addressed in [37e08e5](https://github.com/tenstorrent/tt-metal/pull/46137/commits/37e08e5e509280fa2e43355d774a840940077bba).

File: tt-train/sources/examples/grpo/boolq_training_example.py [hunk 2]
```diff
@@ -41,23 +61,151 @@ def on_step_end(self, trainer, step, **kwargs):
...
+    else:
+        model_id = args.model_source or "meta-llama/Llama-3.2-1B-Instruct"
+        default_config = "tt-train/configs/training_configs/grpo_boolq_llama_1dev.yaml"
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/examples/grpo/boolq_training_example.py]: Addressed in [37e08e5](https://github.com/tenstorrent/tt-metal/pull/46137/commits/37e08e5e509280fa2e43355d774a840940077bba).

File: tt-train/sources/ttml/ttml/trainers/grpo_trainer.py [hunk 3]
```diff
@@ -507,12 +555,27 @@ def train(self) -> None:
...
+            # accumulated since the last reset; without this, the first
+            # training backward would also traverse the old-prob forwards'
+            # nodes. That is harmless for plain models, but FSDP installs
+            # autograd-callback nodes (backward_pre/backward_post) on every
+            # forward, and firing those stale callbacks on already-resharded
+            # modules reduce-scatters/gathers against freed buffers. The
+            # detached old-prob log-prob values live on their tensor objects
+            # (held in probs_old_list) and survive the reset.
+            # ttml.autograd.AutoContext.get_instance().reset_graph()
+            # awliu: reset_graph added in failed attempt to find crash which was actually in loss.backwards()
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/ttml/trainers/grpo_trainer.py]: Fixed in [4ea33e4](https://github.com/tenstorrent/tt-metal/pull/46137/commits/4ea33e4f8d9c7b0cd12bec8fe02b6a4435e1b994).

File: tt-train/sources/examples/grpo/utils/qwen3_completer.py [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+        loss_mask_tt = ttml.autograd.Tensor.from_numpy(
+            loss_mask_np, ttnn.Layout.ROW_MAJOR, ttnn.DataType.BFLOAT16, self._dp_mapper
+        )
+
+        # NOTE: do NOT deallocate ``input_tensor`` / ``mask`` here. When this runs
+        # for the *new* log-probs (grad enabled) they are still referenced by the
+        # autograd graph and are needed by ``loss.backward()``; the trainer's
+        # ``reset_graph()`` releases them after the backward pass.
+
+        # awliu: deleted in failed attempt to find crash which was actually in loss.backwards()
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/examples/grpo/utils/qwen3_completer.py]: Fixed in [4ea33e4](https://github.com/tenstorrent/tt-metal/pull/46137/commits/4ea33e4f8d9c7b0cd12bec8fe02b6a4435e1b994).

File: tt-train/sources/examples/grpo/boolq_training_example.py [hunk 5]
```diff
@@ -13,16 +13,32 @@
...
 from ttml.common.utils import get_tt_metal_runtime_root
 from ttml.trainers import GRPOTrainer, TrainerCallback, get_grpo_config
 from utils.llama_completer import LlamaCompletionCtx
 from utils.llama_completer import LlamaGRPOCompleter
+from utils.qwen3_completer import Qwen3CompletionCtx
+from utils.qwen3_completer import Qwen3GRPOCompleter
+
+import random
+import numpy as np
+import torch
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/examples/grpo/boolq_training_example.py]: Fixed in [4ea33e4](https://github.com/tenstorrent/tt-metal/pull/46137/commits/4ea33e4f8d9c7b0cd12bec8fe02b6a4435e1b994).

File: tt-train/sources/examples/grpo/utils/qwen3_support.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    the topology is unavailable (e.g. unit mesh). Mirrors
+    ``ttml.fsdp._get_placements``; kept local so this module doesn't depend on
+    an underscore-private symbol from ``ttml.fsdp``.
+    """
+    try:
+        value = param_tensor.get_value()
+        topology = value.tensor_topology()
+        return list(topology.placements())
+    except Exception:
+        return None
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/examples/grpo/utils/qwen3_support.py]: Outdated

File: tt-train/sources/examples/grpo/utils/qwen3_completer.py [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+        import torch
+
+        if os.path.isdir(model_source):
+            path = model_source
+        else:
+            path = snapshot_download(
+                repo_id=model_source,
+                allow_patterns=["*.safetensors", "*.json", "*.model", "*.txt"],
+            )
+        hf_model = AutoModelForCausalLM.from_pretrained(path, torch_dtype=torch.float32, trust_remote_code=True)
```
Comment (awliu-TT) [inline tt-train/sources/examples/grpo/utils/qwen3_completer.py:206]: Note about here: This materializes the entire model in host RAM as float32. We then extract the torch state dict, as this is what the ttml qwen3 `load_weights_from_hf` expects as input. This is simple logic, but would be a point to change if we are ever bottlenecked by host RAM. The Llama completer reads raw `.safetensors` diretly and never instantiates a torch model.

File: tt-train/sources/examples/grpo/utils/qwen3_completer.py [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+        # tokens at [0, len_b), pad after). Per-row prediction at len_b - 1.
+        Np = round_up_to_tile(max_prompt_len)
+        pred_pos = [length - 1 for length in lengths]
+
+        tokens_to_complete = min(ctx.max_tokens_to_complete, self._max_seq_len - Np)
+        tokens_to_complete = max(tokens_to_complete, 0)
+
+        stop_ids = self._get_stop_ids()
+
+        print(
```
Comment (jmalone-tt) [inline tt-train/sources/examples/grpo/utils/qwen3_completer.py]: Logging instead of prints?

File: tt-train/sources/examples/grpo/boolq_training_example.py [hunk 9]
```diff
@@ -13,16 +13,27 @@
...
-from ttml.common.config import DeviceConfig, TrainingConfig, get_model_config, load_config
+from ttml.common.config import DeviceConfig, TrainingConfig, TransformerConfig, get_model_config, load_config
 from ttml.common.utils import get_tt_metal_runtime_root
 from ttml.trainers import GRPOTrainer, TrainerCallback, get_grpo_config
 from utils.llama_completer import LlamaCompletionCtx
 from utils.llama_completer import LlamaGRPOCompleter
+from utils.qwen3_completer import Qwen3CompletionCtx
+from utils.qwen3_completer import Qwen3GRPOCompleter
+
+try:
```
Comment (jmalone-tt) [inline tt-train/sources/examples/grpo/boolq_training_example.py:25]: Is there a reason to have the super conservative import instead of just failing?

**Review by jmalone-tt — APPROVED** [review]
Small notes, then looks good to me

**Review by abustamanteTT — APPROVED** [review]

File: tt-train/sources/examples/grpo/boolq_training_example.py (hunk 9 above)
Comment (awliu-TT) [reply tt-train/sources/examples/grpo/boolq_training_example.py:25]: I believe WandB is very useful for experiment logging/tracking, but it's not something that strictly needs to be used. I don't want to prevent people from running the example just because they have issues with WandB. I will add a logging warning when the import fails to notify the user.

File: tt-train/sources/examples/grpo/utils/qwen3_completer.py (hunk 8 above)
Comment (awliu-TT) [reply tt-train/sources/examples/grpo/utils/qwen3_completer.py]: Done in [61573b2](https://github.com/tenstorrent/tt-metal/pull/46137/commits/61573b20380001b562596c661d12af2d8b2999ca).
